lib/thrift/generator/struct_binary_protocol.ex

```diff
@@ -803,6 +771,90 @@ defmodule Thrift.Generator.StructBinaryProtocol do
     |> list_deserializer(name, file_group)
   end
 
+  def struct_serializer(%Union{fields: fields}, name, file_group) do
+    fields = Enum.reject(fields, &(&1.type == :void))
+
+    single_field_serializers = Enum.map(fields, fn
+      %Field{name: match_name} = match_field ->
+        nil_matchers = Enum.flat_map(fields, fn
+          %Field{name: ^match_name} ->
+            []
+          %Field{name: nil_name} ->
+            [{nil_name, nil}]
+        end)
+        field_matchers = [{match_name, Macro.var(match_name, nil)} | nil_matchers]
+
+        field_serializers = [required_field_serializer(match_field, file_group)]
+
+        quote do
+          def serialize(%unquote(name){unquote_splicing(field_matchers)}) do
+            unquote([field_serializers, <<0>>] |> Utils.merge_binaries)
+          end
+        end
+    end)
+
+    all_fields_nil = Enum.map(fields, fn
```

Comment:
If you need to build this list regardless, you can also do this step earlier and then quickly filter it to produce `nil_matchers` using just:

```elixir
nil_matchers = Enum.reject(all_fields_nil, fn({nil_name, _}) -> nil_name == name end)
```